### PR TITLE
Improve toFixed function

### LINKED ISSUE #1367: "Number.prototype.toFixed()" is not properly implemented
The current implementation fails with following test case:
- `assert((0.995).toFixed(2) === "0.99");`
- `assert((9.995).toFixed(2) === "9.99");`

Reference:
- http://www.ecma-international.org/ecma-262/5.1/#sec-15.7.4.5

### PR DESCRIPTION
**Pull request by rerobika**
Fixes #1367.
From now numbers are represented as binary floating-point which guarantees the expected operation of toFixed function.

**Benchmark result** 

Benchmark | RSS (bytes) | Perf (sec) |
----: | ----: | ----: | 
3d-cube.js | 57344 -> 57344 : 0.000% | 0.868 -> 0.863 : +0.615% | 
3d-raytrace.js | 126976 -> 120539 : +5.069% | 1.060 -> 1.049 : +0.994% | 
access-binary-trees.js | 53248 -> 53248 : 0.000% | 0.561 -> 0.565 : -0.573% | 
access-fannkuch.js | 20480 -> 20480 : 0.000% | 2.192 -> 2.184 : +0.327% | 
access-nbody.js | 28672 -> 28672 : 0.000% | 1.105 -> 1.102 : +0.268% | 
bitops-3bit-bits-in-byte.js | 16384 -> 16384 : 0.000% | 0.584 -> 0.583 : +0.106% | 
bitops-bits-in-byte.js | 16384 -> 16384 : 0.000% | 0.863 -> 0.864 : -0.100% | 
bitops-bitwise-and.js | 16384 -> 16384 : 0.000% | 1.007 -> 1.006 : +0.066% | 
bitops-nsieve-bits.js | 114688 -> 114688 : 0.000% | 1.399 -> 1.380 : +1.380% | 
controlflow-recursive.js | 86016 -> 86016 : 0.000% | 0.374 -> 0.371 : +0.886% | 
crypto-aes.js | 69046 -> 69046 : 0.000% | 0.959 -> 0.958 : +0.159% | 
crypto-md5.js | 131072 -> 131072 : 0.000% | 0.663 -> 0.659 : +0.613% | 
crypto-sha1.js | 86016 -> 86016 : 0.000% | 0.649 -> 0.652 : -0.463% | 
date-format-tofte.js | 40960 -> 40960 : 0.000% | 0.746 -> 0.746 : -0.026% | 
date-format-xparb.js | 40960 -> 40960 : 0.000% | 0.411 -> 0.411 : -0.003% | 
math-cordic.js | 16384 -> 16384 : 0.000% | 1.306 -> 1.291 : +1.159% | 
math-partial-sums.js | 16384 -> 16384 : 0.000% | 0.749 -> 0.751 : -0.195% | 
math-spectral-norm.js | 24576 -> 24576 : 0.000% | 0.582 -> 0.584 : -0.198% | 
string-base64.js | 143360 -> 143360 : 0.000% | 1.670 -> 1.541 : +7.740% | 
string-fasta.js | 32768 -> 32768 : 0.000% | 1.210 -> 1.212 : -0.155% | 
Geometric mean: | +0.260% | +0.645% | 

Binary sizes (bytes)
bb8ac29cce:140208
90166803af:140208

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c [hunk 1]
```diff
@@ -509,7 +545,7 @@ ecma_builtin_number_prototype_object_to_fixed (ecma_value_t this_arg, /**< this
       bool is_negative = false;
       if (ecma_number_is_negative (this_num))
       {
-        is_negative = true;
+        is_negative = ecma_number_is_zero (this_num) ? false : true;
```
Comment (robertsipka) [inline jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c]: nice catch, it follows the ECMA-262 standards.

File: jerry-core/ecma/base/ecma-helpers-errol.c [hunk 2]
```diff
@@ -230,6 +230,113 @@ ecma_errol0_dtoa (double val, /**< ecma number */
...
+    mod_res = fmod (val, divider);
+    double residual = mod_res - prev_residual;
+    buffer_p[num_of_digits - 1 - char_cnt++] = (lit_utf8_byte_t) ((int) (residual / (divider / 10.0)) + '0');
+
+    divider *= 10.0;
+    prev_residual = mod_res;
+  }
+
+  *exp_p = char_cnt;
+
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: unnecessary empty line

File: jerry-core/ecma/base/ecma-helpers-errol.c [hunk 3]
```diff
@@ -230,6 +230,113 @@ ecma_errol0_dtoa (double val, /**< ecma number */
...
+} /* ecma_double_to_ascii */
+
+/**
+ * Errol0 double to binary floating-point number conversion
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: Is this part of the original algorithm? https://github.com/marcandrysco/Errol

File: jerry-core/ecma/base/ecma-helpers-conversion.c [hunk 4]
```diff
@@ -931,6 +931,28 @@ ecma_number_to_decimal (ecma_number_t num, /**< ecma-number */
...
+  *
+  * Note:
+  *      The calculated values correspond to s, n, k parameters in ECMA-262 v5, 9.8.1, item 5:
+  *         - parameter out_digits_p corresponds to s, the digits of the number;
+  *         - parameter out_decimal_exp_p corresponds to n, the decimal exponent;
+  *         - return value corresponds to k, the number of digits.
+  */
+lit_utf8_size_t
+ecma_number_to_binary_floating_point_number (ecma_number_t num, /**< ecma-number */
+                       lit_utf8_byte_t *out_digits_p, /**< [out] buffer to fill with digits */
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-conversion.c]: Please align this line correctly.

File: jerry-core/ecma/base/ecma-helpers-errol.c [hunk 5]
```diff
@@ -230,6 +230,113 @@ ecma_errol0_dtoa (double val, /**< ecma number */
...
+ecma_number_of_digits (double val) /**< ecma number */
+{
+  JERRY_ASSERT (val > 0.0);
+  JERRY_ASSERT (fabs (fmod (val, 1.0)) < ERROL0_EPSILON);
+  int32_t exponent = 0;
+
+  while (val >= 1.0)
+  {
+    val /= 10.0;
+    exponent++;
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: Dividing by 10 can be slow if the exponent is large.

File: jerry-core/ecma/base/ecma-helpers-errol.c [hunk 6]
```diff
@@ -230,6 +230,113 @@ ecma_errol0_dtoa (double val, /**< ecma number */
...
+  }
+
+  return exponent;
+} /* ecma_number_of_digits */
+
+inline static void __attr_always_inline___
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: Missing description.

File: jerry-core/ecma/base/ecma-helpers-errol.c [hunk 7]
```diff
@@ -230,6 +230,113 @@ ecma_errol0_dtoa (double val, /**< ecma number */
...
+ * Errol0 double to binary floating-point number conversion
+ *
+ * @return number of generated digits
+ */
+inline lit_utf8_size_t __attr_always_inline___
+ecma_errol0_dtobfpn (double val, /**< ecma number */
+                     lit_utf8_byte_t *buffer_p, /**< buffer to generate digits into */
+                     int32_t *exp_p) /**< [out] exponent */
+{
+  lit_utf8_byte_t integer_part_buffer[325];
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: That is a quite a lot of stack. Why this number is used btw?

File: jerry-core/ecma/base/ecma-helpers-errol.c [hunk 8]
```diff
@@ -230,6 +230,113 @@ ecma_errol0_dtoa (double val, /**< ecma number */
...
+  else
+  {
+    ecma_double_to_ascii (val, buffer_p, &char_cnt);
+  }
+
+  *exp_p = char_cnt;
+
+  while (fraction_part > 0 && char_cnt < ECMA_MAX_CHARS_IN_STRINGIFIED_NUMBER - 1)
+  {
+    fraction_part*=10;
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: Spaces aroung `*=`

File: jerry-core/ecma/base/ecma-helpers.h [hunk 9]
```diff
@@ -355,6 +358,7 @@ lit_utf8_size_t ecma_number_to_utf8_string (ecma_number_t num, lit_utf8_byte_t *
 
 /* ecma-helpers-errol.c */
 lit_utf8_size_t ecma_errol0_dtoa (double val, lit_utf8_byte_t *buffer_p, int32_t *exp_p);
+lit_utf8_size_t ecma_errol0_dtobfpn (double val, lit_utf8_byte_t *buffer_p, int32_t *exp_p);
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers.h]: What does dtobfpn mean?

File: tests/jerry/number-prototype-to-fixed.js [hunk 10]
```diff
@@ -31,14 +31,16 @@ assert((-Infinity).toFixed(0) === "-Infinity");
 assert(NaN.toFixed(0) === "NaN");
 assert((0.0).toFixed(0) === "0");
 assert((0.0).toFixed(1) === "0.0");
-assert((-0.0).toFixed(0) === "-0");
```
Comment (zherczeg) [inline tests/jerry/number-prototype-to-fixed.js]: Why is this changed?

File: jerry-core/ecma/base/ecma-helpers-errol.c (hunk 7 above)
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: Good catch. Much less is enough! IEEE 754 double precision floating point range is beetween ± ≈10^−323.3 to ≈10^308.3

File: tests/jerry/number-prototype-to-fixed.js (hunk 10 above)
Comment (rerobika) [inline tests/jerry/number-prototype-to-fixed.js]: According to **http://www.ecma-international.org/ecma-262/5.1/#sec-15.7.4.5**

```
If x < 0, then
Let s be "-".
Let x = –x.
```
And `-0.0 < 0.0` gives false because they are equal.

File: jerry-core/ecma/base/ecma-helpers-errol.c (hunk 3 above)
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers-errol.c]: No it isn't. I've renamed and moved it to `ecma-helpers-conversion.c`.

File: tests/jerry/number-prototype-to-fixed.js (hunk 10 above)
Comment (akosthekiss) [inline tests/jerry/number-prototype-to-fixed.js]: Note: both Safari and Firefox dev consoles report `(-0.0).toFixed(0) === "-0"` as being `false`

**Comment by glistening** [comment]
@rerobika Could you let me know how did you measure binary size? When I build on x86, I got 221,896 bytes for `jerry` executable.

**Comment by rerobika** [comment]
I've used @bzsolt 's script, I think he can give more detailed information about how it works than I could.

**Comment by glistening** [comment]
@rerobika Could you check binary size again? I think binary size will be increased since new functions are added.

When I measured by `jerry` built with

$ ./tools/build.py --toolchain=cmake/toolchain_linux_armv7l.cmake --clean

Binary sizes (bytes)
5b22e5e:145889
5b22e5e + this patch: 148457

Binary size is increased by 2568 bytes.

**Comment by bzsolt** [comment]
@glistening these results based on the **jerry-minimal** (`main-unix-minimal.c`) instead of the default **jerry** (`main-unix.c`) commandline tool.
To build this binary you have to pass the `--jerry-cmdline-minimal=on` option to `tools/build.py`.
I've rebased this PR to `5b22e5e`, the used compiler is gcc-4.9.2.

| | jerry-minimal | jerry |
| ---- | ---- | ---- |
| master (5b22e5e) | 140208 | 144336 |
| pr_1924 | 140208 | 144384 |

**Comment by glistening** [comment]
@bzsolt @rerobika I've checked again with `--jerry-cmdline-minimal=on`. There is about 700 bytes increment on my build with `gcc-4.8.2.'. I think it is acceptable increment than 2.5 KB that I saw in default build.

2a2a15f: 140773
2a2a15f+pr_1924: 141477

```
$ arm-linux-gnueabihf-gcc -v
Using built-in specs.
COLLECT_GCC=arm-linux-gnueabihf-gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc-cross/arm-linux-gnueabihf/4.8/lto-wrapper
Target: arm-linux-gnueabihf
Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro 4.8.2-16ubuntu4' --with-bugurl=file:///usr/share/doc/gcc-4.8/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.8 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/arm-linux-gnueabihf/include/c++/4.8.2 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libmudflap --disable-libitm --disable-libquadmath --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.8-armhf-cross/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.8-armhf-cross --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.8-armhf-cross --with-arch-directory=arm --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --disable-libgcj --enable-objc-gc --enable-multiarch --enable-multilib --disable-sjlj-exceptions --with-arch=armv7-a --with-fpu=vfpv3-d16 --with-float=hard --with-mode=thumb --disable-werror --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=arm-linux-gnueabihf --program-prefix=arm-linux-gnueabihf- --includedir=/usr/arm-linux-gnueabihf/include
Thread model: posix
gcc version 4.8.2 (Ubuntu/Linaro 4.8.2-16ubuntu4) 
```

I also checked the binary size on iotjs binary. It seems ok. Thank you for improvement.

**Comment by yichoi** [comment]
LGTM
